fluent/src/langneg.js

```diff
@@ -0,0 +1,246 @@
+ * with several modifications:
+ *
+ *  1) available locales are treated as ranges
+ *
+ *    This change allows us to match a more specific request against
+ *    more generic available locale.
+ *
+ *    For example, if the available locale list provides locale `en`,
+ *    and the requested locale is `en-US`, we treat the available locale as
+ *    a locale that matches all possible english requests.
+ *
+ *    This means that we expect available locale ID to be as precize as
+ *    the matches they want to cover.
+ *
+ *    For example, if there is only `sr` available, it's ok to list
+ *    it in available locales. But once the available locales has both,
+ *    Cyrl and Latn variants, the locale IDs should be `sr-Cyrl` and `sr-Latn`
+ *    to avoid any `sr-*` request to match against whole `sr` range.
+ *
+ *    What it does ([requested] * [available] = [supported]):
+ *
+ *    ['en-US'] * ['en'] = ['en']
+ *
+ *  2) likely subtags from LDML 4.3 Likely Subtags has been added
+ *
+ *    The most obvious likely subtag that can be computed is a duplication
+ *    of the language field onto region field (`fr` => `fr-FR`).
+ *
+ *    On top of that, likely subtags may use a list of mappings, that
+ *    allow the algorithm to handle non-obvious matches.
+ *    For example, making sure that we match `en` to `en-US` or `sr` to
+ *    `sr-Cyrl`, while `sr-RU` to `sr-Latn-RU`.
+ *
+ *    This list can be taken directly from CLDR Supplemental Data.
+ *
+ *    What it does ([requested] * [available] = [supported]):
+ *
+ *    ['fr'] * ['fr-FR'] = ['fr-FR']
+ *    ['en'] * ['en-US'] = ['en-US']
+ *    ['sr'] * ['sr-Latn', 'sr-Cyrl'] = ['sr-Cyrl']
+ *
+ *  3) region range check has been added
+ *
+ *    Lastly, the last form of check is against the requested locale ID
+ *    but with the region field replaced with a `*` range.
+ *
+ *    The rationale here laid out in LDML 4.4 Language Matching:
+ *      "(...) normally the fall-off between the user's languages is
+ *      substantially greated than regional variants."
+ *
+ *    In other words, if we can't match for the given region, maybe
+ *    we can match for the same language/script but other region, and
+ *    it will in most cases be preferred over falling back on the next
+ *    language.
+ *
+ *    What it does ([requested] * [available] = [supported]):
+ *
+ *    ['en-AU'] * ['en-US'] = ['en-US']
+ *    ['sr-RU'] * ['sr-Latn-RO'] = ['sr-Latn-RO'] // sr-RU -> sr-Latn-RU
+ */
+function filterMatches(requestedLocales, availableLocales) {
+  const supportedLocales = [];
+
+  outer:
+  for (const requestedLocale of requestedLocales) {
+
+    for (const availableLocale of availableLocales) {
+      if (requestedLocale === availableLocale) {
+        supportedLocales.push(availableLocale);
+        continue outer;
+      }
+    }
+
+    for (const availableLocale of availableLocales) {
+      if (compareLocales(requestedLocale, availableLocale)) {
```

Comment:
This parses the language tags over and over in a loop, I'd refactor this code to not do that.